app/services/role_checker.rb

```diff
@@ -0,0 +1,49 @@
+class RoleChecker
+  attr_reader :user, :records, :viewer_project_ids
+
+  EDITOR_ROLES = %w(owner collaborator expert scientist moderator)
+  APPROVER_ROLES = %w(owner collaborator)
+  VIEWER_ROLES = %w(owner collaborator expert scientist tester)
+
+  def initialize(user, records)
+    @user = user
+    @records = records
+    @viewer_project_ids = get_viewer_project_ids
+  end
+
+  def can_edit?
+    ids = user_role_ids(user.roles, EDITOR_ROLES)
+    check_roles(ids, records)
+  end
+
+  def can_approve?
+    ids = user_role_ids(user.roles, APPROVER_ROLES)
+    check_roles(ids, records)
+  end
+
+  def can_view?
+    ids = user_role_ids(user.roles, VIEWER_ROLES)
+    check_roles(ids, records)
+  end
+
+  def get_viewer_project_ids
+    user_role_ids(user.roles, VIEWER_ROLES)
+  end
+
+  def user_role_ids(user_roles, allowed_roles)
```

Comment:
I feel that the name of this method doesn't make it clear that it's returning project ids. Maybe a comment above the method (or a renaming of the method name - e.g. `user_project_ids` ?) could help clarify this?